README.rst
*****************
webargs-starlette
*****************

.. image:: https://badgen.net/pypi/v/webargs-starlette
    :target: https://badge.fury.io/py/webargs-starlette
    :alt: PyPI version

.. image:: https://dev.azure.com/sloria/sloria/_apis/build/status/sloria.webargs-starlette?branchName=master
    :target: https://dev.azure.com/sloria/sloria/_build/latest?definitionId=11&branchName=master
    :alt: Build status

.. image:: https://badgen.net/badge/marshmallow/3
    :target: https://marshmallow.readthedocs.io/en/stable/
    :alt: marshmallow 3 compatible

.. image:: https://badgen.net/badge/code%20style/black/000
    :target: https://github.com/ambv/black
    :alt: code style: black


webargs-starlette is a library for declarative request parsing and
validation with `Starlette <https://github.com/encode/starlette>`_,
built on top of `webargs <https://github.com/marshmallow-code/webargs>`_.

It has all the goodness of `webargs <https://github.com/marshmallow-code/webargs>`_, 
with some extra sugar for type annotations.

.. code-block:: python

    import uvicorn
    from starlette.applications import Starlette
    from starlette.responses import JSONResponse
    from webargs_starlette import use_annotations

    app = Starlette()


    @app.route("/")
    @use_annotations(location="query")
    async def index(request, name: str = "World"):
        return JSONResponse({"Hello": name})


    if __name__ == "__main__":
        uvicorn.run(app, port=5000)

    # curl 'http://localhost:5000/'
    # {"Hello": "World"}
    # curl 'http://localhost:5000/?name=Ada'
    # {"Hello": "Ada"}

Install
=======

::

    pip install -U webargs-starlette


Usage
=====

Parser Usage
------------

Use ``parser.parse`` to parse a Starlette ``Request`` with a
dictionary of fields.

.. code-block:: python

    from starlette.applications import Starlette
    from starlette.responses import JSONResponse
    from webargs import fields
    from webargs_starlette import parser

    app = Starlette()


    @app.route("/")
    async def homepage(request):
        args = {
            "name": fields.Str(required=True),
            "greeting": fields.Str(load_default="hello"),
        }
        parsed = await parser.parse(args, request)
        greeting = parsed["greeting"]
        name = parsed["name"]
        return JSONResponse({"message": f"{greeting} {name}"})


Decorators
----------

Use the ``use_args`` decorator to inject the parsed arguments
dictionary into the handler function. The following snippet is equivalent to the
first example.

**Important**: Decorated functions MUST be coroutine functions.

.. code-block:: python

    from starlette.applications import Starlette
    from starlette.responses import JSONResponse
    from webargs import fields
    from webargs_starlette import use_args

    app = Starlette()


    @app.route("/")
    @use_args(
        {"name": fields.Str(required=True), "greeting": fields.Str(load_default="hello")}
    )
    async def homepage(request, args):
        greeting = args["greeting"]
        name = args["name"]
        return JSONResponse({"message": f"{greeting} {name}"})


The ``use_kwargs`` decorator injects the parsed arguments as keyword arguments.

.. code-block:: python

    from starlette.applications import Starlette
    from starlette.responses import JSONResponse
    from webargs import fields
    from webargs_starlette import use_args

    app = Starlette()


    @app.route("/")
    @use_kwargs(
        {"name": fields.Str(required=True), "greeting": fields.Str(load_default="hello")}
    )
    async def homepage(request, name, greeting):
        return JSONResponse({"message": f"{greeting} {name}"})


See `decorator_example.py <https://github.com/sloria/webargs-starlette/blob/master/examples/decorator_example.py>`_
for a more complete example of ``use_args`` and ``use_kwargs`` usage.

Error Handling
--------------

When validation fails, the parser will raise a ``WebargsHTTPException``,
which is the same as Starlette's ``HTTPException`` with the addition of
of the ``messages`` (validation messages), ``headers`` , ``exception`` (underlying exception), and ``schema`` (marshmallow ``Schema``) attributes.

You can use a custom exception handler to return the error messages as
JSON.


.. code-block:: python

    from starlette.responses import JSONResponse
    from webargs_starlette import WebargsHTTPException


    @app.exception_handler(WebargsHTTPException)
    async def http_exception(request, exc):
        return JSONResponse(exc.messages, status_code=exc.status_code, headers=exc.headers)


Annotations
-----------

The ``use_annotations`` decorator allows you to parse request objects
using type annotations.


.. code-block:: python

    from starlette.applications import Starlette
    from starlette.responses import JSONResponse
    from webargs_starlette import use_annotations

    app = Starlette()


    @app.route("/")
    @use_annotations(location="query")
    async def welcome(request, name: str = "Friend"):
        return JSONResponse({"message": f"Welcome, {name}!"})


    # curl 'http://localhost:5000/'.
    # {"message":"Welcome, Friend!"}
    # curl 'http://localhost:5000/?name=Ada'.
    # {"message":"Welcome, Ada!"}

Any annotated argument that doesn't have a default value will be required.
For example, if we remove the default for ``name`` in the above example,
an 422 error response is returned if ``?name`` isn't passed.


.. code-block:: python

    from starlette.applications import Starlette
    from starlette.responses import JSONResponse
    from webargs_starlette import use_annotations, WebargsHTTPException

    app = Starlette()


    @app.route("/")
    @use_annotations(location="query")
    async def welcome(request, name: str):
        return JSONResponse({"message": f"Welcome, {name}!"})


    @app.exception_handler(WebargsHTTPException)
    async def http_exception(request, exc):
        return JSONResponse(exc.messages, status_code=exc.status_code, headers=exc.headers)


    # curl "http://localhost:5000/"
    # {"name":["Missing data for required field."]}

Arguments may also be annotated with ``Field`` instances when you need
more control. For example, you may want to add a validator.

.. code-block:: python

    from starlette.applications import Starlette
    from starlette.responses import JSONResponse
    from webargs import fields
    from marshmallow import validate
    from webargs_starlette import use_annotations, WebargsHTTPException

    app = Starlette()


    @app.route("/")
    @use_annotations(location="query")
    async def welcome(request, name: fields.Str(validate=validate.Length(min=2))):
        return JSONResponse({"message": f"Welcome, {name}!"})


    @app.exception_handler(WebargsHTTPException)
    async def http_exception(request, exc):
        return JSONResponse(exc.messages, status_code=exc.status_code, headers=exc.headers)


    # curl "http://localhost:5000/?name=A"
    # {"name":["Shorter than minimum length 2."]}

``HTTPEndpoint`` methods may also be decorated with ``use_annotations``.

.. code-block:: python

    from starlette.applications import Starlette
    from starlette.responses import JSONResponse
    from starlette.endpoints import HTTPEndpoint
    from webargs_starlette import use_annotations

    app = Starlette()


    @app.route("/")
    class WelcomeEndpoint(HTTPEndpoint):
        @use_annotations(location="query")
        async def get(self, request, name: str = "World"):
            return JSONResponse({"message": f"Welcome, {name}!"})

See `annotation_example.py <https://github.com/sloria/webargs-starlette/blob/master/examples/annotation_example.py>`_
for a more complete example of ``use_annotations`` usage.

More
----

For more information on how to use webargs, see the `webargs documentation <https://webargs.readthedocs.io/>`_.

License
=======

MIT licensed. See the `LICENSE <https://github.com/sloria/webargs-starlette/blob/master/LICENSE>`_ file for more details.


CHANGELOG.rst
*********
Changelog
*********

2.1.0 (2022-12-04)
------------------

Features:

* Tested against Python 3.9, 3.10, and 3.11.

Bug fixes:

* Various typing fixes.

Other changes:

* Drop support for Python 3.6.

Thanks `@DmitryBurnaev <https://github.com/DmitryBurnaev>`_ for the PR for these changes.

2.0.0 (2020-05-05)
------------------

Features:

* *Backwards-incompatible*: Support webargs 6.x. webargs<6 is no longer supported.
* Tested against Python 3.8.

Other changes:

* *Backwards-incompatible*: Drop support for marshmallow 2.x. Only marshmallow>=3 is supported.

1.2.1 (2020-05-04)
------------------

Bug fixes:

* Handle ``JSONDecodeError`` when parsing JSON.

1.2.0 (2019-07-05)
------------------

Features:

* Export type annotations.

1.1.0 (2019-01-13)
------------------

Features:

* ``use_annotations`` supports more types in the ``typing`` module, as
  well as ``dict`` and ``list``.
* ``use_annotations`` supports ``typing.Optional``.
* Improve type annotations.

1.0.0 (2019-01-04)
------------------

* First release.


tests/app.py
from starlette.applications import Starlette
from starlette.responses import JSONResponse as J
from starlette.endpoints import HTTPEndpoint
import marshmallow as ma
from webargs import fields
from webargs_starlette import (
    parser,
    use_args,
    use_kwargs,
    use_annotations,
    WebargsHTTPException,
)

app = Starlette()
app.debug = True


class HelloSchema(ma.Schema):
    name = fields.Str(load_default="World", validate=lambda n: len(n) >= 3)


hello_args = {"name": fields.Str(load_default="World", validate=lambda n: len(n) >= 3)}
hello_multiple = {"name": fields.List(fields.Str())}
hello_exclude_schema = HelloSchema(unknown=ma.EXCLUDE)
hello_many_schema = HelloSchema(many=True)


@app.route("/echo", methods=["GET"])
async def echo(request):
    return J(await parser.parse(hello_args, request, location="query"))


@app.route("/echo_query")
async def echo_query(request):
    parsed = await parser.parse(hello_args, request, location="query")
    return J(parsed)


@app.route("/echo_json", methods=["POST"])
async def echo_json(request):
    parsed = await parser.parse(hello_args, request, location="json")
    return J(parsed)


@app.route("/echo_form", methods=["POST"])
async def echo_form(request):
    parsed = await parser.parse(hello_args, request, location="form")
    return J(parsed)


@app.route("/echo_json_or_form", methods=["POST"])
async def echo_json_or_form(request):
    parsed = await parser.parse(hello_args, request, location="json_or_form")
    return J(parsed)


@app.route("/echo_ignoring_extra_data", methods=["POST"])
async def echo_json_ignore_extra_data(request):
    parsed = await parser.parse(hello_exclude_schema, request)
    return J(parsed)


@app.route("/echo_use_args", methods=["GET"])
@use_args(hello_args, location="query")
async def echo_use_args(request, args):
    return J(args)


@app.route("/echo_use_args_validated", methods=["POST"])
@use_args(
    {"value": fields.Int()}, validate=lambda args: args["value"] > 42, location="form"
)
async def echo_use_args_validated(request, args):
    return J(args)


@app.route("/echo_use_kwargs", methods=["GET"])
@use_kwargs(hello_args, location="query")
async def echo_use_kwargs(request, name):
    return J({"name": name})


@app.route("/echo_multi", methods=["GET"])
async def echo_multi(request):
    result = await parser.parse(hello_multiple, request, location="query")
    return J(result)


@app.route("/echo_multi_form", methods=["POST"])
async def echo_multi_form(request):
    result = await parser.parse(hello_multiple, request, location="form")
    return J(result)


@app.route("/echo_multi_json", methods=["POST"])
async def echo_multi_json(request):
    result = await parser.parse(hello_multiple, request, location="json")
    return J(result)


@app.route("/echo_many_schema", methods=["GET", "POST"])
async def many_nested(request):
    arguments = await parser.parse(hello_many_schema, request, location="json")
    return J(arguments)


@app.route("/echo_use_args_with_path_param/{name}")
@use_args({"value": fields.Int()}, location="query")
async def echo_use_args_with_path(request, args):
    return J(args)


@app.route("/echo_use_kwargs_with_path_param/{name}")
@use_kwargs({"value": fields.Int()}, location="query")
async def echo_use_kwargs_with_path(request, value):
    return J({"value": value})


@app.route("/error", methods=["GET", "POST"])
async def error(request):
    def always_fail(value):
        raise ma.ValidationError("something went wrong")

    args = {"text": fields.Str(validate=always_fail)}
    return J(await parser.parse(args, request))


@app.route("/echo_headers")
async def echo_headers(request):
    # the "exclude schema" must be used in this case because WSGI headers may
    # be populated with many fields not sent by the caller
    return J(await parser.parse(hello_exclude_schema, request, location="headers"))


@app.route("/echo_cookie")
async def echo_cookie(request):
    return J(await parser.parse(hello_args, request, location="cookies"))


@app.route("/echo_nested", methods=["POST"])
async def echo_nested(request):
    args = {"name": fields.Nested({"first": fields.Str(), "last": fields.Str()})}
    parsed = await parser.parse(args, request)
    return J(parsed)


@app.route("/echo_nested_many", methods=["POST"])
async def echo_nested_many(request):
    args = {
        "users": fields.Nested({"id": fields.Int(), "name": fields.Str()}, many=True)
    }
    parsed = await parser.parse(args, request)
    return J(parsed)


@app.route("/echo_path_param/{path_param}")
async def echo_path_param(request):
    parsed = await parser.parse(
        {"path_param": fields.Int()}, request, location="path_params"
    )
    return J(parsed)


@app.route("/echo_endpoint/")
class EchoEndpoint(HTTPEndpoint):
    @use_args(hello_args, location="query")
    async def get(self, request, args):
        return J(args)


@app.route("/echo_endpoint_annotations/")
class EchoEndpointAnnotations(HTTPEndpoint):
    @use_annotations(location="query")
    async def get(self, request, name: str = "World"):
        return J({"name": name})


@app.exception_handler(WebargsHTTPException)
async def http_exception(request, exc):
    return J(exc.messages, status_code=exc.status_code, headers=exc.headers)


tests/test_example.py
import pytest

from starlette.testclient import TestClient

from examples.starlette_example import app


@pytest.fixture()
def client():
    return TestClient(app)


def test_index(client):
    assert client.get("/").json() == {"message": "Welcome, Friend!"}
    assert client.get("/?name=Ada").json() == {"message": "Welcome, Ada!"}


def test_add(client):
    assert client.post("/add", json={"x": 2, "y": 2}).json() == {"result": 4}


def test_dateadd(client):
    assert client.post(
        "/dateadd", json={"value": "2019-01-03", "addend": 1}
    ).json() == {"result": "2019-01-04"}


def test_error(client):
    res = client.post("/dateadd", json={})
    assert res.status_code == 422
    assert res.json() == {"addend": ["Missing data for required field."]}

    res = client.post("/dateadd", json={"addend": "invalid"})
    assert res.status_code == 422
    assert "addend" in res.json()

    res = client.post("/dateadd", json={"addend": 42, "unit": "invalid"})
    assert res.status_code == 422
    assert "unit" in res.json()


tests/webtest_asgi.py
import typing

import webob
import webtest
from starlette.testclient import TestClient
from starlette.types import ASGIApp


WSGIApp = typing.Callable[
    [typing.MutableMapping, typing.Callable], typing.List[typing.ByteString]
]


def AsgiToWsgi(asgi_app: ASGIApp) -> WSGIApp:
    """
    This is copy-paste AsgiToWsgi from webtest-asgi library,
    but adopted for new version for httpx-based starlette's TestClient
    """

    def handle(environ, start_response):
        """Support logic and avoiding deprecation warnings for httpx-based TestClient"""
        req = webob.Request(environ)
        with TestClient(asgi_app, cookies=dict(req.cookies)) as client:
            content_kwargs = {}
            # sending raw-bytes content must be through "content" argument
            # avoiding DeprecationWarning: Use 'content=<...>' to upload raw bytes/text content.
            if req.body is not None and not isinstance(req.body, dict):
                content_kwargs["content"] = req.body
            else:
                content_kwargs["data"] = req.body

            response = client.request(
                method=req.method,
                url=req.url,
                headers=dict(req.headers),
                **content_kwargs,
            )

        res = webob.Response(
            body=response.content,
            status=response.status_code,
            content_type=response.headers.get("content-type"),
            headerlist=list(response.headers.items()),
            charset=response.encoding,
        )
        start_response(res.status, res.headerlist)
        return res.app_iter

    return handle


class WebTestApp(webtest.TestApp):
    """A modified `webtest.TestApp` that can wrap an ASGI application. Takes the same
    arguments as `webtest.TestApp`.
    """

    def __init__(self, app: ASGIApp, *args, **kwargs) -> None:
        self.asgi_app = app
        super().__init__(AsgiToWsgi(app), *args, **kwargs)


examples/decorator_example.py
"""A simple number and datetime addition JSON API.
Uses ``use_args`` and ``use_kwargs`` for parsing requests.

Run the app:

    $ pip install uvicorn
    $ python examples/decorator_example.py

Or, to run with automatic reloading:

    $ uvicorn examples.starlette_example:app --port 5001 --debug

Try the following with httpie (a cURL-like utility, http://httpie.org):

    $ pip install httpie
    $ http GET :5001/
    $ http GET :5001/ name==Ada
    $ http POST :5001/add x=40 y=2
    $ http POST :5001/dateadd value=1973-04-10 addend=63
    $ http POST :5001/dateadd value=2014-10-23 addend=525600 unit=minutes
"""
import datetime as dt

from starlette.applications import Starlette
from starlette.responses import JSONResponse
from webargs import fields, validate
from webargs_starlette import use_args, use_kwargs, WebargsHTTPException

app = Starlette()


@app.route("/")
@use_args({"name": fields.Str(load_default="Friend")}, location="querystring")
async def index(request, args):
    """A welcome page."""
    return JSONResponse({"message": f"Welcome, {args['name']}!"})


@app.route("/add", methods=["POST"])
@use_kwargs(
    {"x": fields.Float(required=True), "y": fields.Float(required=True)},
    location="json",
)
async def add(request, x, y):
    """An addition endpoint."""
    return JSONResponse({"result": x + y})


@app.route("/dateadd", methods=["POST"])
@use_kwargs(
    {
        "value": fields.Date(required=False),
        "addend": fields.Int(required=True, validate=validate.Range(min=1)),
        "unit": fields.Str(
            load_default="days", validate=validate.OneOf(["minutes", "days"])
        ),
    },
    location="json",
)
async def dateadd(request, value, addend, unit):
    """A datetime adder endpoint."""
    value = value or dt.datetime.utcnow()
    if unit == "minutes":
        delta = dt.timedelta(minutes=addend)
    else:
        delta = dt.timedelta(days=addend)
    result = value + delta
    return JSONResponse({"result": result.isoformat()})


# Return errors as JSON
@app.exception_handler(WebargsHTTPException)
async def http_exception(request, exc):
    return JSONResponse(exc.messages, status_code=exc.status_code, headers=exc.headers)


if __name__ == "__main__":
    import uvicorn

    app.debug = True
    uvicorn.run(app, host="0.0.0.0", port=5001)


examples/starlette_example.py
"""A simple number and datetime addition JSON API.
Run the app:

    $ python examples/starlette_example.py

Or, to run with automatic reloading:

    $ uvicorn examples.starlette_example:app --port 5001 --debug

Try the following with httpie (a cURL-like utility, http://httpie.org):

    $ pip install httpie
    $ http GET :5001/
    $ http GET :5001/ name==Ada
    $ http POST :5001/add x=40 y=2
    $ http POST :5001/dateadd value=1973-04-10 addend=63
    $ http POST :5001/dateadd value=2014-10-23 addend=525600 unit=minutes
"""
import uvicorn
import datetime as dt

from starlette.applications import Starlette
from starlette.responses import JSONResponse
from webargs import fields, validate
from webargs_starlette import use_args, use_kwargs, WebargsHTTPException

app = Starlette()


@app.route("/")
@use_args({"name": fields.Str(missing="Friend")})
async def index(request, args):
    """A welcome page."""
    return JSONResponse({"message": f"Welcome, {args['name']}!"})


@app.route("/add", "POST")
@use_kwargs({"x": fields.Float(required=True), "y": fields.Float(required=True)})
async def add(request, x, y):
    """An addition endpoint."""
    return JSONResponse({"result": x + y})


@app.route("/dateadd", "POST")
@use_kwargs(
    {
        "value": fields.Date(required=False),
        "addend": fields.Int(required=True, validate=validate.Range(min=1)),
        "unit": fields.Str(
            missing="days", validate=validate.OneOf(["minutes", "days"])
        ),
    }
)
async def dateadd(request, value, addend, unit):
    """A datetime adder endpoint."""
    value = value or dt.datetime.utcnow()
    if unit == "minutes":
        delta = dt.timedelta(minutes=addend)
    else:
        delta = dt.timedelta(days=addend)
    result = value + delta
    return JSONResponse({"result": result.isoformat()})


# Return errors as JSON
@app.exception_handler(WebargsHTTPException)
async def http_exception(request, exc):
    return JSONResponse(exc.messages, status_code=exc.status_code, headers=exc.headers)


if __name__ == "__main__":
    app.debug = True
    uvicorn.run(app, host="0.0.0.0", port=5001)


webargs_starlette/annotations.py
import typing
import inspect
from collections import abc

from starlette.requests import Request
from marshmallow import Schema, fields
from marshmallow.fields import Field


DEFAULT_TYPE_MAPPING = Schema.TYPE_MAPPING.copy()
DEFAULT_TYPE_MAPPING.update(
    {
        dict: fields.Dict,
        list: fields.List,
        abc.Mapping: fields.Dict,
        typing.Mapping: fields.Dict,
        typing.Dict: fields.Dict,
        abc.Iterable: fields.List,
        typing.Iterable: fields.List,
        abc.Sequence: fields.List,
        typing.Sequence: fields.List,
        abc.MutableSequence: fields.List,
        typing.MutableSequence: fields.List,
        typing.List: fields.List,
        typing.AnyStr: fields.String,
    }
)


# Marshmallow type mapping
TypeMapping = typing.Mapping[typing.Type, typing.Type[Field]]


class TypeMappingError(TypeError):
    def __init__(self, param: str, cls: type) -> None:
        super().__init__(
            "Cannot create field from type annotation "
            f'for "{param}". Pass a TypeMapping '
            f"that includes {cls}."
        )


def _type2field(
    name: str,
    type_: type,
    signature: inspect.Signature,
    type_mapping: TypeMapping,
    **kwargs,
) -> Field:
    if isinstance(type_, Field):
        return type_
    else:
        default = signature.parameters[name].default
        required = default is inspect.Parameter.empty
        field_kwargs = {"required": required}
        args = getattr(type_, "__args__", [])

        if not required:
            field_kwargs["load_default"] = default

        origin_cls = getattr(type_, "__origin__", None) or type_
        try:
            field_cls = type_mapping[origin_cls]
        except KeyError as key_err:
            if type(type_) is typing.TypeVar:
                field_cls = fields.Field
            # typing.Optional[T] or typing.Union[T, None] -> fields.Field(allow_none=True)
            elif origin_cls is typing.Union and type(None) in args:
                field_kwargs["allow_none"] = True
                non_none_args = [
                    arg for arg in args if arg is not type(None)  # noqa: E721
                ]
                # If only one other type is passed, get the proper field for that type
                # e.g. typing.Union[int, None] -> fields.Int(allow_none=True)
                if len(non_none_args) == 1:
                    try:
                        field_cls = type_mapping[non_none_args[0]]
                    except KeyError as err:
                        raise TypeMappingError(name, origin_cls) from err
                else:
                    field_cls = fields.Field
            else:
                raise TypeMappingError(name, origin_cls) from key_err

        # Handle container fields
        if issubclass(field_cls, fields.List):
            args = getattr(type_, "__args__", [])
            if args:
                inner_type = args[0]
                container = _type2field(name, inner_type, signature, type_mapping)
            else:
                container = fields.Field()
            field_kwargs["cls_or_instance"] = container
        elif issubclass(field_cls, fields.Dict):
            if args:
                key_type, val_type = args
                key_container = _type2field(name, key_type, signature, type_mapping)
                field_kwargs["keys"] = key_container
                value_container = _type2field(name, val_type, signature, type_mapping)
                field_kwargs["values"] = value_container
        field_kwargs.update(kwargs)
        return field_cls(**field_kwargs)


def annotations2schema(
    func: typing.Callable, type_mapping: typing.Optional[TypeMapping] = None
) -> typing.Type[Schema]:
    type_mapping = type_mapping or DEFAULT_TYPE_MAPPING
    annotations = getattr(func, "__annotations__", {})
    signature = inspect.signature(func)
    fields_dict = {}
    for name, annotation in annotations.items():
        # Skip over request argument and return annotation
        if name == "return" or (
            isinstance(annotation, type) and issubclass(annotation, Request)
        ):
            continue

        fields_dict[name] = _type2field(name, annotation, signature, type_mapping)

    schema_class = Schema.from_dict(fields_dict)
    return typing.cast(typing.Type[Schema], schema_class)


webargs_starlette/starletteparser.py
import typing
import functools
import json

from marshmallow import Schema, ValidationError
from starlette.requests import Request
from starlette.exceptions import HTTPException
from starlette.endpoints import HTTPEndpoint
from webargs.asyncparser import AsyncParser
from webargs import core
from webargs.multidictproxy import MultiDictProxy
from .annotations import TypeMapping, DEFAULT_TYPE_MAPPING, annotations2schema

HTTP_METHOD_NAMES: typing.List[str] = [
    "get",
    "post",
    "put",
    "patch",
    "delete",
    "head",
    "options",
    "trace",
]


class WebargsHTTPException(HTTPException):
    """
    Same as `starlette.exceptions.HTTPException` but stores validation
    messages, the underlying exception, the `marshmallow.Schema`
    used to parse the request, headers.
    """

    def __init__(
        self,
        status_code: int,
        detail: typing.Optional[str] = None,
        messages: typing.Optional[dict] = None,
        exception: typing.Optional[Exception] = None,
        headers: typing.Optional[dict] = None,
        schema: typing.Optional[Schema] = None,
    ) -> None:
        super().__init__(status_code, detail)
        self.messages = messages
        self.exception = exception
        self.headers = headers
        self.schema = schema


def is_json_request(req: Request) -> bool:
    content_type = req.headers.get("content-type")
    return core.is_json(content_type)


class StarletteParser(AsyncParser):
    """Starlette request argument parser."""

    TYPE_MAPPING: TypeMapping = DEFAULT_TYPE_MAPPING

    __location_map__: typing.Dict[str, typing.Union[str, typing.Callable]] = dict(
        path_params="load_path_params", **core.Parser.__location_map__
    )

    def load_path_params(self, req: Request, schema: Schema) -> typing.Any:
        """Return the request's ``path_params`` or ``missing`` if there are none."""
        return req.path_params or core.missing

    def load_querystring(self, req: Request, schema: Schema) -> MultiDictProxy:
        """Return query params from the request as a MultiDictProxy."""
        return MultiDictProxy(req.query_params, schema)

    def load_headers(self, req: Request, schema: Schema) -> MultiDictProxy:
        """Return headers from the request as a MultiDictProxy."""
        return MultiDictProxy(req.headers, schema)

    def load_cookies(self, req: Request, schema: Schema):
        """Return cookies from the request."""
        return req.cookies

    async def load_json(self, req: Request, schema: Schema) -> typing.Dict:
        """Return a parsed json payload from the request."""
        if not is_json_request(req):
            return core.missing
        try:
            json_data = await req.json()
        except json.JSONDecodeError as exc:
            if exc.doc == "":
                return core.missing
            else:
                return self._handle_invalid_json_error(exc, req)
        except UnicodeDecodeError as exc:
            return self._handle_invalid_json_error(exc, req)
        return json_data

    async def load_form(self, req: Request, schema: Schema) -> MultiDictProxy:
        """Return form values from the request as a MultiDictProxy."""
        post_data = await req.form()
        return MultiDictProxy(post_data, schema)

    async def load_json_or_form(
        self, req: Request, schema: Schema
    ) -> typing.Union[typing.Dict, MultiDictProxy]:
        data = await self.load_json(req, schema)
        if data is not core.missing:
            return data
        return await self.load_form(req, schema)

    def _handle_invalid_json_error(
        self, error: Exception, req: Request, *args, **kwargs
    ) -> typing.NoReturn:
        raise WebargsHTTPException(
            400, exception=error, messages={"json": ["Invalid JSON body."]}
        )

    def get_request_from_view_args(
        self, view: typing.Callable, args: tuple, kwargs: dict
    ) -> Request:
        """Get request object from a handler function or method. Used internally by
        ``use_args`` and ``use_kwargs``.
        """
        req = None
        for arg in args:
            if isinstance(arg, Request):
                req = arg
                break
        assert isinstance(req, Request), "Request argument not found for handler"
        return req

    def handle_error(
        self,
        error: ValidationError,
        req: Request,
        schema: Schema,
        error_status_code: typing.Union[int, None],
        error_headers: typing.Union[dict, None],
    ) -> typing.NoReturn:
        """Handles errors during parsing. Aborts the current HTTP request and
        responds with a 422 error.
        """
        status_code = error_status_code or self.DEFAULT_VALIDATION_STATUS
        raise WebargsHTTPException(
            status_code,
            exception=error,
            messages=error.messages,
            schema=schema,
            headers=error_headers,
        )

    def use_annotations(
        self,
        fn: typing.Union[typing.Callable, typing.Type[HTTPEndpoint], None] = None,
        *,
        type_mapping: TypeMapping = None,
        **kwargs,
    ) -> typing.Union[typing.Callable[..., typing.Callable], typing.Type[HTTPEndpoint]]:
        # Allow using this as either a decorator or a decorator factory.
        if fn is None:
            return functools.partial(
                use_annotations, type_mapping=type_mapping, **kwargs
            )
        type_mapping = type_mapping or self.TYPE_MAPPING

        def decorator(func: typing.Callable) -> typing.Callable:
            schema = annotations2schema(func, type_mapping=type_mapping)

            @functools.wraps(func)
            async def wrapper(*a, **kw):
                request = self.get_request_from_view_args(func, a, kw)
                parsed = await self.parse(schema, request, **kwargs)
                kw.update(parsed)
                return await func(*a, **kw)

            return wrapper

        if isinstance(fn, type) and issubclass(fn, HTTPEndpoint):
            endpoint = fn
            # Decorate each HTTP-mapped method
            for each in HTTP_METHOD_NAMES:
                if hasattr(endpoint, each):
                    handler = getattr(endpoint, each)
                    setattr(endpoint, each, decorator(handler))
            return endpoint
        else:
            return decorator(fn)


parser = StarletteParser()
use_args = parser.use_args
use_kwargs = parser.use_kwargs
use_annotations = parser.use_annotations


examples/annotation_example.py
"""A simple number and datetime addition JSON API.
Uses ``use_annotations`` for parsing requests.

Run the app:

    $ pip install uvicorn
    $ python examples/annotation_example.py

Or, to run with automatic reloading:

    $ uvicorn examples.annotation_example:app --port 5001 --debug

Try the following with httpie (a cURL-like utility, http://httpie.org):

    $ pip install httpie
    $ http GET :5001/
    $ http GET :5001/ name==Ada
    $ http POST :5001/add x=40 y=2
    $ http POST :5001/dateadd value=1973-04-10 addend=63
    $ http POST :5001/dateadd value=2014-10-23 addend=525600 unit=minutes
"""
import typing
import datetime as dt

from starlette.applications import Starlette
from starlette.requests import Request
from starlette.responses import JSONResponse, Response
from starlette.endpoints import HTTPEndpoint
from webargs import fields, validate
from webargs_starlette import use_annotations, WebargsHTTPException

app = Starlette()


@app.route("/")
@use_annotations(location="query")
async def welcome(request: Request, name: str = "Friend") -> Response:
    """A welcome page."""
    return JSONResponse({"message": f"Welcome, {name}!"})


@typing.no_type_check
@app.route("/welcome2")
@use_annotations(location="query")
async def welcome2(
    request: Request,
    name: fields.Str(load_default="Friend", metadata={"location": "querystring"}),
) -> Response:
    """A welcome page, using a field annotation."""
    return JSONResponse({"message": f"Welcome, {name}!"})


@app.route("/welcome3")
@use_annotations(location="query")
async def welcome_no_default(request: Request, name: str) -> Response:
    """A welcome page with no default name. If "name" isn't passed in the querystring,
    an error response will be returned.
    """
    return JSONResponse({"message": f"Welcome, {name}!"})


@app.route("/add", methods=["POST"])
@use_annotations(location="json")
async def add(request: Request, x: float, y: float) -> Response:
    """An addition endpoint."""
    return JSONResponse({"result": x + y})


@app.route("/add2")
class AddEndpoint(HTTPEndpoint):
    """An addition endpoint, implemented as a class."""

    @use_annotations(location="json")
    async def post(self, request: Request, x: float, y: float) -> Response:
        return JSONResponse({"result": x + y})


@typing.no_type_check
@app.route("/dateadd", methods=["POST"])
@use_annotations(location="json")
async def dateadd(
    request: Request,
    addend: fields.Int(required=True, validate=validate.Range(min=1)),
    unit: fields.Str(load_default="days", validate=validate.OneOf(["minutes", "days"])),
    value: dt.date = None,
) -> Response:
    """A datetime adder endpoint."""
    value = value or dt.datetime.utcnow()
    if unit == "minutes":
        delta = dt.timedelta(minutes=addend)
    else:
        delta = dt.timedelta(days=addend)
    result = value + delta
    return JSONResponse({"result": result.isoformat()})


# Return errors as JSON
@app.exception_handler(WebargsHTTPException)
async def http_exception(request: Request, exc: WebargsHTTPException) -> Response:
    return JSONResponse(exc.messages, status_code=exc.status_code, headers=exc.headers)


if __name__ == "__main__":
    import uvicorn

    app.debug = True
    uvicorn.run(app, host="0.0.0.0", port=5001)
